Make sure that DensityEnergyFromPressureTemperature works for all equations of state

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "fix bug in ideal gas EOS.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

The easiest way to get the Jacobian for the PT-space PTE solver we are building is to ensure that the `DensityEnergyFromPressureTemperature` call works consistently for all equations of state. This MR makes sure thats the case and threads this call through all the tests. 

To ease this process, I added a base class implementation. However, I think most classes will want their own implementations, and I added them when appropriate.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.
- [ ] Ensure that any `when='@main'` dependencies are updated to the release version in the package.py

Yurlungur commented on .github/workflows/tests_minimal.yml:
```diff
@@ -36,4 +36,4 @@ jobs:
             make -j4
             make install
-            make test
+            ctest --output-on-failure
```
I think I'd like to leave this permanently. You still get a summary out at the end, but here we can see what went wrong with a test. Useful for debugging, especially when the CI machine and your laptop disagree...

Yurlungur commented on singularity-eos/eos/eos_base.hpp:
```diff
@@ -836,6 +838,44 @@ class EosBase {
     PORTABLE_ALWAYS_THROW_OR_ABORT(msg);
   }
 
+  // JMM: This method is often going to be overloaded for special cases.
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp,
+                                       Indexer_t &&lambda, Real &rho, Real &sie) const {
```
This default implementation is usually not optimal but does seem to work if a given EOS doesn't need special cases. In general, though, an EOS should overload and implement its own.

Yurlungur commented on singularity-eos/eos/eos_base.hpp:
```diff
@@ -836,6 +838,44 @@ class EosBase {
+      }
+    }
+    auto status = findRoot(PofRT, press, rhoguess, rhomin, rhomax, EPS, EPS, rho);
+    // JMM: This needs to not fail and instead return something sane
+    if (status != Status::SUCCESS) {
+      PORTABLE_WARN("DensityEnergyFromPressureTemperature failed to find root\n");
+      rho = rhoguess;
+    }
```
This is a function that may be passed data and I see it as similar to a PTE solve. Even if it fails out, something reasonable must be done.

Yurlungur commented on singularity-eos/eos/eos_helmholtz.hpp:
```diff
@@ -676,16 +676,37 @@ class Helmholtz : public EosBase<Helmholtz> {
-    // reference state, maybe stellar denity, would be appropriate.
-    PORTABLE_ALWAYS_ABORT("Stub");
+    // TODO(JMM): Is this right???
+    rho = 1e7;
+    temp = 1.5e9;
```
Conditions for an oxygen burning shell in a stellar core.

jhp-lanl replied:
Can  you comment this in the code?

Yurlungur replied:
:+1: done

Yurlungur commented on singularity-eos/eos/eos_mgusup.hpp:
```diff
@@ -125,6 +125,47 @@ class MGUsup : public EosBase<MGUsup> {
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real MinimumDensity() const { return 10 * robust::EPS(); }
+
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real MaximumDensity() const {
+    if (_s > 1) {
+      return 0.99 * robust::ratio(_s * _rho0, _s - 1);
+    } else { // for s <= 1, no maximum, but we need to pick something.
+      return 1e3 * _rho0;
+    }
+  }
```
These are the conditions where the reference isotherm is a positive temperature.

jhp-lanl replied:
Honestly the $s < 0$ case implies that the fundamental shock derivative is negative and the EOS is no longer convex, so it's not a very useful EOS. I think what you picked is as good as any for that case.

Yurlungur replied:
:+1: I will add a comment to this effect.

Yurlungur commented on singularity-eos/eos/eos_mgusup.hpp:
```diff
@@ -125,6 +125,47 @@ class MGUsup : public EosBase<MGUsup> {
           const unsigned long output,
           Indexer_t &&lambda = static_cast<Real *>(nullptr)) const;
+
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real MinimumDensity() const { return 10 * robust::EPS(); }
```
Since the reference isotherm scales as _rho0/rho, the EOS diverges at zero density. This bounds it at least to some degree.

Yurlungur commented on singularity-eos/eos/eos_stellar_collapse.hpp:
```diff
@@ -653,7 +653,25 @@ PORTABLE_INLINE_FUNCTION Real StellarCollapse::GruneisenParamFromDensityInternal
+  if ((lrguess < lRhoMin_) || (lrguess > lRhoMax_)) {
+    lrguess = lRho_(rhoNormal_);
+  }
+  auto lPofRT = [&](Real lR) { return lP_.interpToReal(Ye, lT, lR); };
+  auto status = regula_falsi(lPofRT, lP, lrguess, lRhoMin_, lRhoMax_, ROOT_THRESH,
+                             ROOT_THRESH, lrguess);
```
Since the EOS is tabulated in logrho, use logrho for the root find.

Yurlungur commented on singularity-eos/eos/eos_vinet.hpp:
```diff
@@ -131,6 +131,10 @@ class Vinet : public EosBase<Vinet> {
                          Indexer_t &&lambda = static_cast<Real *>(nullptr)) const;
+
+  PORTABLE_FORCEINLINE_FUNCTION
+  Real MinimumDensity() const { return std::cbrt(10 * robust::EPS()) * _rho0; }
```
EOS depends on $\left(\frac{\rho_0}{\rho}\right)^{1/3}$ so a reasonable bound before the EOS will become ill behaved is this.

Yurlungur commented on singularity-eos/eos/eos_vinet.hpp:
```diff
@@ -152,11 +156,6 @@ class Vinet : public EosBase<Vinet> {
     }
     printf("\n\n");
   }
-  // Density/Energy from P/T not unique, if used will give error
-  template <typename Indexer_t = Real *>
-  PORTABLE_INLINE_FUNCTION void
-  DensityEnergyFromPressureTemperature(const Real press, const Real temp,
-                                       Indexer_t &&lambda, Real &rho, Real &sie) const;
```
use the default in the base class

Yurlungur commented on test/eos_unit_test_helpers.hpp:
```diff
@@ -143,6 +144,34 @@ inline void compare_two_eoss(const E1 &test_e, const E2 &ref_e) {
   return;
 }
 
+template <typename EOS, typename Indexer_t = Real *>
+PORTABLE_INLINE_FUNCTION bool
+CheckRhoSieFromPT(EOS eos, Real rho, Real T,
+                  Indexer_t &&lambda = static_cast<Real *>(nullptr)) {
```
Convenience function to let me test the functionality more easily

Yurlungur commented:
Ok that was finicky but tests on github now seem to be passing. re-git tests appear to be having some network issues. Ping @rbberger for after the holidays. :)

jhp-lanl commented on singularity-eos/eos/eos_base.hpp:
```diff
@@ -836,6 +838,44 @@ class EosBase {
+    // overwrite.
+    constexpr Real MAXFAC = 1e8;
+    constexpr Real EPS = 10 * robust::EPS();
+    constexpr Real MINR_DEFAULT = 1e-4;
```
Is this a meaningful value? If not, I'd pick something more like 1e-08. That _should_ be low enough for most sesame tables.

Yurlungur replied:
It's not meaningful. I think I can actually make these hardcoded constants go away by being a bit more clever.

jhp-lanl commented on singularity-eos/eos/eos_mgusup.hpp:
```diff
@@ -125,6 +125,47 @@ class MGUsup : public EosBase<MGUsup> {
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp,
```
Why not just use the base class version?

Yurlungur replied:
Because of the maximum density, which is a runtime parameter for eos_mgusup but not the others. In the future we might want introspection for max density though. I'll add a comment explaining and a TODO to remove this when max density is available.

Yurlungur replied:
Actually, we should just provide a max density that can be overwritten. I will add that.

jhp-lanl commented on singularity-eos/eos/eos_mgusup.hpp:
```diff
@@ -200,6 +236,8 @@ PORTABLE_INLINE_FUNCTION Real MGUsup::HugTemperatureFromDensity(Real rho) const
   Real eta = 1.0 - robust::ratio(_rho0, rho);
   Real f1 = 1.0 - _s * eta;
   if (f1 <= 0.0) {
+    printf("f1, eta, rho, rho0, s = %.14e %.14e %.14e %.14e %.14e\n", f1, eta, rho, _rho0,
+           _s);
```
Good addition!

Yurlungur commented:
OK... that took longer than I thought it would. All tests now passing once again and the machinery is more flexible and robust. @jhp-lanl please re-review when you get the chance. I believe I've addressed all your concerns. 

I also noticed that the `DensityEnergyFromPressureTemperature` function wasn't documented. It's documented now.

Yurlungur commented:
Ping @jhp-lanl

jhp-lanl commented:
> Ping @jhp-lanl

I'll try to re-review that by the end of today.

Yurlungur commented:
Snuck in introspection for pressure into this MR as well.

jhp-lanl reviewed (approved):
Approving because none of my comments are really blocking. That said, I do think the name change to `MaximumPressureAtTemperature()` would be preferable. I also think the documentation should be changed to indicate that there _could_ be a temperature dependence for the maximum pressure if an EOS supports it.

Yurlungur commented on doc/sphinx/src/using-eos.rst:
```diff
@@ -1143,18 +1143,75 @@ quantities as outputs.
+provides the minimum pressure an equation of state supports, which may
+be the most negative tension state. The function
+
+.. cpp:function:: Real MaximumPressureFromTemperature() const;
+
+provides a maximum possible pressure an equation of state
+supports. (Most models are unbounded in pressure.) This is again
+useful for root finds.
```
I completely agree--I change the name. For the tabulated equations of state, I actually do pull it out and record it now (thinking ahead), but I let this float as the table interpolation does extrapolate off the top of the table. It's not particularly trustworthy, but it's not a hard limit. Only Gruneisen currently reports a maximum pressure bound. On the other hand, I do have each table report the minimum pressure, which is important.

Yurlungur commented on doc/sphinx/src/using-eos.rst:
```diff
@@ -1143,18 +1143,75 @@ quantities as outputs.
+
+.. note::
+
+  Note that ``lambda`` must be passed in, whether or not a given
+  equation of state requires one. You may pass in ``nullptr`` safely,
+  however.
```
I was going to say it's because of the order of the arguments---you can't have a default argument in the middle. But you know what? There's no reason we can't just add an overload in the base class. Done.

Yurlungur commented on singularity-eos/eos/eos_eospac.hpp:
Agree this is a concern. I made an issue, #454, but will punt for now.

Yurlungur commented:
Thanks for the review, @jhp-lanl ! Agree with your comments and have addressed them. Will merge once tests pass on re-git.
